packages/protoc-gen-es/src/util.ts

```diff
@@ -0,0 +1,315 @@
+  if (defaultValue === undefined) {
+    return undefined;
+  }
+  switch (field.fieldKind) {
+    case "enum": {
+      const enumValue = field.enum.values.find(
+        (value) => value.number === defaultValue,
+      );
+      if (enumValue === undefined) {
+        throw new Error(
+          `invalid enum default value: ${String(defaultValue)} for ${enumValue}`,
+        );
+      }
+      return literalEnumValue(enumValue, opt);
+    }
+    case "scalar":
+      return literalScalarValue(defaultValue, field, opt.protoInt64Symbol);
+  }
+}
+
+/**
+ * Return a printable expression for the zero value of a field.
+ *
+ * Returns either:
+ * - empty array literal for repeated fields
+ * - empty object literal for maps
+ * - undefined for message fields
+ * - an enums first value
+ * - scalar zero value
+ */
+export function getFieldZeroValueExpression(
+  field: DescField | DescExtension,
+  opt: GetFieldExpressionOptions,
+): Printable | undefined {
+  if (field.repeated) {
+    return "[]";
+  }
+  switch (field.fieldKind) {
+    case "message":
+      return undefined;
+    case "map":
+      return "{}";
+    case "enum": {
+      // In proto3, the first enum value must be zero.
+      // In proto2, protobuf-go returns the first value as the default.
+      if (field.enum.values.length < 1) {
+        throw new Error("invalid enum: missing at least one value");
+      }
+      const zeroValue = field.enum.values[0];
+      return literalEnumValue(zeroValue, opt);
+    }
+    case "scalar": {
+      const defaultValue = codegenInfo.scalarDefaultValue(
+        field.scalar,
+        field.longType,
+      ) as string | boolean | number | bigint | Uint8Array;
+      return literalScalarValue(defaultValue, field, opt.protoInt64Symbol);
+    }
+  }
+}
+
+function literalScalarValue(
+  value: string | number | bigint | boolean | Uint8Array,
+  field: (DescField | DescExtension) & { fieldKind: "scalar" },
+  protoInt64Symbol: ImportSymbol,
+): Printable {
+  switch (field.scalar) {
+    case ScalarType.DOUBLE:
+    case ScalarType.FLOAT:
+    case ScalarType.INT32:
+    case ScalarType.FIXED32:
+    case ScalarType.UINT32:
+    case ScalarType.SFIXED32:
+    case ScalarType.SINT32:
+      if (typeof value != "number") {
+        throw new Error(
+          `Unexpected value for ${ScalarType[field.scalar]} ${field.toString()}: ${String(value)}`,
+        );
+      }
+      return value;
+    case ScalarType.BOOL:
+      if (typeof value != "boolean") {
+        throw new Error(
+          `Unexpected value for ${ScalarType[field.scalar]} ${field.toString()}: ${String(value)}`,
+        );
+      }
+      return value;
+    case ScalarType.STRING:
+      if (typeof value != "string") {
+        throw new Error(
+          `Unexpected value for ${ScalarType[field.scalar]} ${field.toString()}: ${String(value)}`,
+        );
+      }
+      return literalString(value);
+    case ScalarType.BYTES:
+      if (!(value instanceof Uint8Array)) {
```

Comment:
Is it worth avoiding `instanceof` here if possible in light of https://github.com/bufbuild/protobuf-es/issues/713?

Reply:
Uint8Array is a built-in global. It isn't ever imported from a package, and it's not subject to the dual package hazard.